patches/v8/0019-Implement-cross-request-context-promise-resolve-hand.patch

```diff
@@ -0,0 +1,339 @@
++  inline void set_promise_cross_context_resolve_callback(
++      PromiseCrossContextResolveCallback callback);
+   MaybeHandle<JSPromise> RunPromiseCrossContextCallback(Handle<NativeContext> context,
+                                                         Handle<JSPromise> promise);
++  void RunPromiseCrossContextResolveCallback(v8::Isolate* isolate,
++                                             Handle<JSObject> tag,
++                                             Handle<Object> reactions,
++                                             Handle<Object> argument,
++                                             PromiseReaction::Type type);
++
++  inline bool has_promise_context_resolve_callback();
+ 
+ #ifdef V8_ENABLE_WASM_SIMD256_REVEC
+   void set_wasm_revec_verifier_for_test(
+@@ -2764,8 +2774,10 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
+ #endif
+ 
+   Tagged<Object> promise_context_tag_;
+-  PromiseCrossContextCallback promise_cross_context_callback_;
++  PromiseCrossContextCallback promise_cross_context_callback_ = nullptr;
++  PromiseCrossContextResolveCallback promise_cross_context_resolve_callback_ = nullptr;
+   bool in_promise_cross_context_callback_ = false;
++  bool in_promise_cross_context_resolve_callback_ = false;
+ 
+   class PromiseCrossContextCallbackScope;
+ 
+diff --git a/src/objects/js-promise.h b/src/objects/js-promise.h
+index e4c1e42c63e..87bebb70d0d 100644
+--- a/src/objects/js-promise.h
++++ b/src/objects/js-promise.h
+@@ -75,6 +75,11 @@ class JSPromise
+   static_assert(v8::Promise::kFulfilled == 1);
+   static_assert(v8::Promise::kRejected == 2);
+ 
++  static void ContinueTriggerPromiseReactions(Isolate* isolate,
++                                              DirectHandle<Object> reactions,
++                                              DirectHandle<Object> argument,
++                                              PromiseReaction::Type type);
++
+  private:
+   // ES section #sec-triggerpromisereactions
+   static Handle<Object> TriggerPromiseReactions(Isolate* isolate,
+diff --git a/src/objects/objects.cc b/src/objects/objects.cc
+index 76e284e19a2..32ccc84d8f8 100644
+--- a/src/objects/objects.cc
++++ b/src/objects/objects.cc
+@@ -4878,6 +4878,21 @@ Handle<Object> JSPromise::Fulfill(DirectHandle<JSPromise> promise,
+   // 6. Set promise.[[PromiseState]] to "fulfilled".
+   promise->set_status(Promise::kFulfilled);
+ 
++  Tagged<Object> obj = promise->context_tag();
++  bool needs_promise_context_switch =
++      !(obj == Smi::zero() ||
++        obj == isolate->promise_context_tag() ||
++        !isolate->has_promise_context_resolve_callback());
++  if (needs_promise_context_switch) {
++    isolate->RunPromiseCrossContextResolveCallback(
++        reinterpret_cast<v8::Isolate*>(isolate),
++        Handle<JSObject>(Cast<JSObject>(obj), isolate),
++        reactions,
++        value,
++        PromiseReaction::kFulfill);
++    return isolate->factory()->undefined_value();
++  }
```

Comment:
This is the actual meat of the patch. If the promise has no context tag, or the context tag matches, or there is no configured callback, then we'll schedule the continuations synchronously per the default behavior. If the promise tag does not match, then we defer scheduling the continuations by calling out to our embedder-configured callback. This needs to happen in both the resolve and reject paths.